nomad/host_volume_endpoint.go

```diff
@@ -0,0 +1,527 @@
+							return false, nil
+						}
+						if args.NodePool != "" && vol.NodePool != args.NodePool {
+							return false, nil
+						}
+
+						if ns != structs.AllNamespacesSentinel &&
+							vol.Namespace != ns {
+							return false, nil
+						}
+
+						allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRead)
+						return allowVolume(aclObj, ns), nil
+					},
+				},
+			}
+
+			// Set up our output after we have checked the error.
+			var vols []*structs.HostVolumeStub
+
+			// Build the paginator. This includes the function that is
+			// responsible for appending a variable to the variables
+			// stubs slice.
+			paginatorImpl, err := paginator.NewPaginator(iter, tokenizer, filters, args.QueryOptions,
+				func(raw any) error {
+					vol := raw.(*structs.HostVolume)
+					vols = append(vols, vol.Stub())
+					return nil
+				})
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "failed to create result paginator: %v", err)
+			}
+
+			// Calling page populates our output variable stub array as well as
+			// returns the next token.
+			nextToken, err := paginatorImpl.Page()
+			if err != nil {
+				return structs.NewErrRPCCodedf(
+					http.StatusBadRequest, "failed to read result page: %v", err)
+			}
+
+			reply.Volumes = vols
+			reply.NextToken = nextToken
+
+			// Use the index table to populate the query meta as we have no way
+			// of tracking the max index on deletes.
+			return v.srv.setReplyQueryMeta(store, state.TableHostVolumes, &reply.QueryMeta)
+		},
+	}
+
+	return v.srv.blockingRPC(&opts)
+}
+
+func (v *HostVolume) Create(args *structs.HostVolumeCreateRequest, reply *structs.HostVolumeCreateResponse) error {
+
+	authErr := v.srv.Authenticate(v.ctx, args)
+	if done, err := v.srv.forward("HostVolume.Create", args, args, reply); done {
+		return err
+	}
+	v.srv.MeasureRPCRate("host_volume", structs.RateMetricWrite, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "host_volume", "create"}, time.Now())
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeCreate)
+	aclObj, err := v.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+	if !allowVolume(aclObj, args.RequestNamespace()) {
+		return structs.ErrPermissionDenied
+	}
+
+	if len(args.Volumes) == 0 {
+		return fmt.Errorf("missing volume definition")
+	}
+
+	for _, vol := range args.Volumes {
+		if vol.Namespace != "" {
+			if !allowVolume(aclObj, vol.Namespace) {
+				return structs.ErrPermissionDenied
+			}
+		} else {
+			vol.Namespace = args.RequestNamespace()
+		}
+	}
+
+	var mErr *multierror.Error
+
+	// ensure we only try to create valid volumes or make valid updates to
+	// volumes
+	validVols, err := v.validateVolumeUpdates(args.Volumes)
+	if err != nil {
+		mErr = multierror.Append(mErr, err)
+	}
+	if len(validVols) == 0 {
+		return helper.FlattenMultierror(mErr)
+	}
```

Comment:
I would think _any_ invalid volumes would be a bad request, and return (multi)error.  why partially execute their request?

Reply:
That would certainly simplify things. I think we still need to handle multierrors for client RPCs (because those won't be atomic and we'll want to record any volumes that we've successfully created/updated on the client), but we can certainly exit early here.